File: src/3-figures.R
# ==============================================================================
# 3-figures.R
#
# Purpose:
#   Produce publication-ready figures for the earnings event study using
#   ggplot2.
#
# Inputs (from DATA_DIR):
#   regdata.parquet
#   figure-data.parquet
#
# Outputs (to OUTPUT_DIR):
#   ff12_fig.{pdf,png}      Bar chart: SameSign frequency by FF12 industry
#   size_year.{pdf,png}     Multi-line time series: SameSign by size quintile
#   corr_fig.{pdf,png}      Correlation matrix heatmap (corrplot)
#   car_fig.{pdf,png}       Event-study CAR plot over the [-5,+5] day window
#   erc_year.{pdf,png}      Year-by-year ERC with confidence bands
#
# Notes:
#   - PDF for LaTeX, PNG for Word / PowerPoint. Both are saved every run.
#   - Demonstrates: geom_col + coord_flip; multi-line time series with
#     grouped aesthetics; corrplot heatmap; nest_by + broom::tidy for
#     year-by-year coefficient extraction with geom_ribbon for the CIs.
# ==============================================================================


# Setup ------------------------------------------------------------------------

if (!require("pacman")) install.packages("pacman")
pacman::p_load(dotenv, lubridate, glue, arrow, forcats, corrplot,
               fixest, broom,
               tidyverse)

load_dot_env(".env")
data_dir   <- Sys.getenv("DATA_DIR")
output_dir <- Sys.getenv("OUTPUT_DIR")

source("src/utils.R")


# Read in the data from the previous step --------------------------------------

# regdata: one row per firm-quarter, with SUE, BHAR, controls
regdata <- read_parquet(glue("{data_dir}/regdata.parquet"))

# figdata: one row per firm-quarter-offset, for the CAR event window plot
figdata <- read_parquet(glue("{data_dir}/figure-data.parquet"))

nrow(regdata)
nrow(figdata)


# Figure 1: SameSign Frequency by Industry -------------------------------------

# Horizontal bar chart showing how often earnings and sales move in the same
# direction (our key interaction variable, SameSign) within each FF12 industry.
# Higher SameSign rates suggest earnings changes are more often "confirmed" by
# revenue movements — relevant background for the SUE × SameSign regression.
#
# This teaches:
#   - geom_col() for bar charts from summarized data
#   - coord_flip() to make horizontal bars (easier to read industry names)
#   - fct_reorder() to sort bars by value instead of alphabetical order
#   - scales::percent for axis formatting
#   - custom fill color

fig <- regdata |>
  group_by(FF12) |>
  summarize(pct_same_sign = mean(same_sign, na.rm = TRUE)) |>
  # Reorder industries by % SameSign so the chart is sorted, not alphabetical
  mutate(FF12 = fct_reorder(factor(FF12), pct_same_sign)) |>
  ggplot(aes(x = FF12, y = pct_same_sign)) +
  geom_col(fill = "#0051ba") +
  # Fill color = Kansas Blue: https://brand.ku.edu/guidelines/design/color
  scale_y_continuous(name = "Freq. of SameSign (Earnings & Sales Same Direction)",
                     labels = scales::percent) +
  scale_x_discrete(name = "Fama-French Industry") +
  coord_flip() +
  # base_family = "serif" sets font to Times New Roman (or similar serif font)
  theme_bw(base_family = "serif")

# Look at it in R
fig

# For LaTeX output you might want to output to PDF
ggsave(glue("{output_dir}/ff12_fig.pdf"), fig, width = 7, height = 6)

# For Word output you might want to output to an image such as .png
ggsave(glue("{output_dir}/ff12_fig.png"), fig, width = 4.2, height = 3.6)


# Figure 2: SameSign Frequency by Size Quintile Over Time ---------------------

# Multi-line time series showing SameSign frequency by size quintile over time.
# Looking at how SameSign rates differ across firm size and across years gives
# a feel for how much variation our interaction variable has in the panel.
#
# This teaches:
#   - ntile() to create quantile groups
#   - Grouped aesthetics (color + linetype mapped to the same variable)
#   - Giving two scales the same name so they share one legend

fig <- regdata |>
  group_by(fyearq) |>
  # Create size quintiles within each year
  mutate(size_qnt = factor(ntile(mve, 5))) |>
  group_by(fyearq, size_qnt) |>
  summarize(pct_same_sign = mean(same_sign, na.rm = TRUE), .groups = "drop") |>
  filter(!is.na(pct_same_sign)) |>
  ggplot(aes(x = fyearq, y = pct_same_sign, color = size_qnt, linetype = size_qnt)) +
  geom_line() + geom_point() +
  scale_y_continuous(name = "Freq. of SameSign", labels = scales::percent) +
  scale_x_continuous(name = "Year", breaks = seq(1970, 2025, 5)) +
  # If you give these two scales the same name, they share one legend
  scale_color_discrete(name = "Size Quintile") +
  scale_linetype_discrete(name = "Size Quintile") +
  theme_bw(base_family = "serif")

fig

ggsave(glue("{output_dir}/size_year.pdf"), fig, width = 7, height = 6)
ggsave(glue("{output_dir}/size_year.png"), fig, width = 7, height = 6)


# Figure 3: Correlation Matrix -------------------------------------------------

# Heatmap of pairwise correlations using the corrplot package.
# This is NOT ggplot — corrplot uses base R graphics. It's included here
# because correlation matrices are standard in empirical papers and corrplot
# makes them look nice with minimal code.
#
# Note the pdf() / dev.off() pattern: corrplot writes to the active graphics
# device, so we open a PDF device first, draw the plot, then close it.

corrdata <- regdata |>
  select(
    `BHAR`     = bhar,
    `SUE`      = sue,
    `SameSign` = same_sign,
    `LOSS`     = loss,
    `SIZE`     = log_mve
  )

corrdata

correlation <- cor(corrdata, use = "pairwise.complete.obs")
col2 <- colorRampPalette(c('red', 'white', 'blue'))

# Save to PDF
pdf(file = glue("{output_dir}/corr_fig.pdf"))
corrplot(correlation, method = 'square',
         addCoef.col = 'black',
         diag = FALSE,
         tl.col = 'black',
         type = 'full',
         tl.cex = 1,
         tl.srt = 0,
         tl.offset = 1,
         number.cex = 0.7,
         cl.ratio = 0.1,
         cl.pos = "r",
         col = col2(20),
         win.asp = .8)
dev.off()


# Figure 4: Event Study CAR Plot ----------------------------------------------

# The classic event study figure: cumulative abnormal returns over the
# [-5, +5] trading day window around earnings announcements.
#
# We split by SameSign (our interaction variable) to show that the market
# reacts more strongly when earnings and sales move in the same direction.
#
# This teaches:
#   - Working with event-time data (offset on x-axis)
#   - geom_hline() for a zero reference line
#   - Grouping by a categorical variable to compare subsamples
#   - Computing means and confidence intervals by group

# Merge SUE and SameSign from regdata into the figure data, then cross them
# into the four cells our regression estimates: SUE direction × SameSign.
car_data <- figdata |>
  inner_join(
    regdata |> select(permno, rdq, same_sign, sue),
    by = c("permno", "rdq")
  ) |>
  mutate(
    sue_dir = case_when(
      sue > 0 ~ "Positive SUE",
      sue < 0 ~ "Negative SUE",
      TRUE    ~ NA_character_
    ),
    same_sign_lbl = if_else(same_sign == 1L, "Same Sign", "Diff Sign")
  ) |>
  filter(!is.na(sue_dir))

# Average cumulative BHAR at each offset, by SUE direction × SameSign
car_summary <- car_data |>
  group_by(sue_dir, same_sign_lbl, offset) |>
  summarize(
    mean_bhar = mean(bhar_cum, na.rm = TRUE),
    se = sd(bhar_cum, na.rm = TRUE) / sqrt(n()),
    n = n(),
    .groups = "drop"
  )

# Color = SUE direction (substantive sign of the surprise);
# Linetype = SameSign (whether sales confirm earnings).
# If SameSign matters, the gap between solid and dashed within each color
# should widen as we move away from the announcement.
fig <- car_summary |>
  ggplot(aes(x = offset, y = mean_bhar,
             color = sue_dir, linetype = same_sign_lbl,
             group = interaction(sue_dir, same_sign_lbl))) +
  geom_hline(yintercept = 0, linetype = "dashed", color = "gray50") +
  geom_line(linewidth = 0.8) +
  geom_point(size = 2) +
  scale_x_continuous(name = "Trading Days Relative to Announcement",
                     breaks = -5:5) +
  scale_y_continuous(name = "Cumulative BHAR", labels = scales::percent) +
  scale_color_manual(name = "Earnings Surprise",
                     values = c("Positive SUE" = "#0051ba",
                                "Negative SUE" = "#c41230")) +
  scale_linetype_manual(name = "SameSign",
                        values = c("Same Sign" = "solid",
                                   "Diff Sign" = "dashed")) +
  theme_bw(base_family = "serif") +
  theme(legend.position = "bottom")

fig

ggsave(glue("{output_dir}/car_fig.pdf"), fig, width = 7, height = 6)
ggsave(glue("{output_dir}/car_fig.png"), fig, width = 7, height = 6)


# Figure 5: Year-by-Year ERC with Confidence Bands ----------------------------

# This plots the earnings response coefficient (ERC) — the slope of
# BHAR on SUE — estimated separately for each year. Confidence bands
# show the precision of the estimate in each year.
#
# This teaches:
#   - nest_by() to split data into groups and fit models to each
#   - broom::tidy() to extract regression coefficients as a tidy data.frame
#   - geom_ribbon() for confidence interval bands
#   - The concept of time-varying coefficients (how the ERC changes over time)

# We split by SameSign to see if the ERC differs when sales confirm earnings

ercdata <- regdata |>
  # nest_by creates a grouped data frame with a list-column of data
  nest_by(fyearq, same_sign) |>
  # Fit a simple regression of BHAR on SUE within each year × SameSign group
  mutate(
    fit = list(lm(bhar ~ sue, data = data))
  ) |>
  # Extract coefficients with confidence intervals
  reframe(broom::tidy(fit, conf.int = TRUE)) |>
  # Drop the intercept — we only want to plot the SUE coefficient
  filter(term != "(Intercept)")

# can also use this setup to do Fama-Macbeth regressions, etc.
# can also use pmg package for Fama-Macbeth

fig <- ercdata |>
  mutate(same_sign = factor(same_sign, labels = c("Different Sign", "Same Sign"))) |>
  ggplot(aes(x = fyearq, y = estimate)) +
  geom_ribbon(aes(ymin = conf.low, ymax = conf.high, group = same_sign),
              fill = "grey80") +
  geom_line(aes(color = same_sign)) +
  geom_point(aes(color = same_sign)) +
  scale_x_continuous(name = "Year", breaks = seq(1970, 2025, 5)) +
  scale_y_continuous(name = "ERC (Coefficient on SUE)") +
  scale_color_manual(name = "",
                     values = c("Same Sign" = "#0051ba",
                                "Different Sign" = "#c41230")) +
  theme_bw(base_family = "serif") +
  theme(legend.position = "bottom")

fig

ggsave(glue("{output_dir}/erc_year.pdf"), fig, width = 7, height = 6)
ggsave(glue("{output_dir}/erc_year.png"), fig, width = 7, height = 6)

File: src/3-figures.py
# ==============================================================================
# 3-figures.py
#
# Purpose:
#   Produce publication-ready figures for the earnings event study using
#   plotnine (Grammar of Graphics in Python — nearly identical syntax to
#   ggplot2 in R, making the two implementations easy to compare).
#
# Inputs (from DATA_DIR):
#   regdata.parquet
#   figure-data.parquet
#
# Outputs (to OUTPUT_DIR):
#   ff12_fig.{pdf,png}      Bar chart: SameSign frequency by FF12 industry
#   size_year.{pdf,png}     Multi-line time series: SameSign by size quintile
#   corr_fig.{pdf,png}      Correlation matrix heatmap (seaborn)
#   car_fig.{pdf,png}       Event-study CAR plot, four lines (SUE × SameSign)
#   erc_year.{pdf,png}      Year-by-year ERC with confidence bands
#
# Notes:
#   - Run via `uv run src/3-figures.py`.
#   - PDF for LaTeX, PNG for Word / PowerPoint. Both are saved every run.
# ==============================================================================


# Setup ------------------------------------------------------------------------

import os
import sys

import numpy as np
import polars as pl
import statsmodels.api as sm
from plotnine import *
from dotenv import load_dotenv

sys.stdout.reconfigure(encoding="utf-8")

load_dotenv(".env", override=True)
data_dir = os.getenv("DATA_DIR")
output_dir = os.getenv("OUTPUT_DIR")

# Ensure output directory exists
os.makedirs(output_dir, exist_ok=True)


# Read in the data from the previous step --------------------------------------

# regdata: one row per firm-quarter, with SUE, BHAR, controls
regdata = pl.read_parquet(f"{data_dir}/regdata.parquet")

# figdata: one row per firm-quarter-offset, for the CAR event window plot
figdata = pl.read_parquet(f"{data_dir}/figure-data.parquet")

print(f"regdata: {regdata.shape[0]:,} rows")
print(f"figdata: {figdata.shape[0]:,} rows")


# Figure 1: SameSign Frequency by Industry -------------------------------------

# Horizontal bar chart showing how often earnings and sales move in the same
# direction (SameSign) within each FF12 industry. Higher rates suggest
# earnings changes are more often "confirmed" by revenue movements.

fig1_data = (regdata
    .group_by("FF12")
    .agg(pl.col("same_sign").mean().alias("pct_same_sign"))
    .sort("pct_same_sign")
    # plotnine respects categorical ordering for the axis
    .with_columns(
        pl.col("FF12").cast(pl.Categorical).alias("FF12")
    )
    .to_pandas()
)

# Reorder the categories by pct_same_sign
import pandas as pd
fig1_data["FF12"] = pd.Categorical(
    fig1_data["FF12"],
    categories=fig1_data.sort_values("pct_same_sign")["FF12"],
    ordered=True,
)

fig1 = (
    ggplot(fig1_data, aes(x="FF12", y="pct_same_sign"))
    + geom_col(fill="#0051ba")
    + coord_flip()
    + scale_y_continuous(
        name="Freq. of SameSign (Earnings & Sales Same Direction)",
        labels=lambda x: [f"{v:.0%}" for v in x],
    )
    + scale_x_discrete(name="Fama-French Industry")
    + theme_bw(base_family="serif")
)

fig1.save(f"{output_dir}/ff12_fig.pdf", width=7, height=6)
fig1.save(f"{output_dir}/ff12_fig.png", width=4.2, height=3.6, dpi=150)
print("Figure 1 saved")


# Figure 2: SameSign Frequency by Size Quintile Over Time ----------------------

# Multi-line time series showing SameSign frequency by size quintile over time.

fig2_data = (regdata
    .with_columns(
        pl.col("mve")
        .qcut(5, labels=["1", "2", "3", "4", "5"])
        .over("fyearq")
        .cast(pl.String)
        .alias("size_qnt")
    )
    .filter(pl.col("size_qnt").is_not_null())
    .group_by("fyearq", "size_qnt")
    .agg(pl.col("same_sign").mean().alias("pct_same_sign"))
    .sort("fyearq", "size_qnt")
    .to_pandas()
)

fig2 = (
    ggplot(fig2_data, aes(x="fyearq", y="pct_same_sign", color="size_qnt"))
    + geom_line()
    + geom_point()
    + scale_y_continuous(
        name="Freq. of SameSign",
        labels=lambda x: [f"{v:.0%}" for v in x],
    )
    + scale_x_continuous(name="Year", breaks=range(1970, 2026, 5))
    + scale_color_discrete(name="Size Quintile")
    + theme_bw(base_family="serif")
)

fig2.save(f"{output_dir}/size_year.pdf", width=7, height=6)
fig2.save(f"{output_dir}/size_year.png", width=7, height=6, dpi=150)
print("Figure 2 saved")


# Figure 3: Correlation Matrix -------------------------------------------------

# Heatmap of pairwise correlations. We use seaborn here (not plotnine)
# because correlation matrix heatmaps are cleaner with seaborn's heatmap().

import matplotlib.pyplot as plt
import seaborn as sns

corrdata = (regdata
    .select(
        pl.col("bhar").alias("BHAR"),
        pl.col("sue").alias("SUE"),
        pl.col("same_sign").alias("SameSign"),
        pl.col("loss").alias("LOSS"),
        pl.col("log_mve").alias("SIZE"),
    )
    .to_pandas()
)

correlation = corrdata.corr()

fig3, ax = plt.subplots(figsize=(7, 6))
sns.heatmap(
    correlation,
    annot=True,
    fmt=".2f",
    cmap=sns.diverging_palette(0, 240, as_cmap=True),
    center=0,
    square=True,
    linewidths=0.5,
    ax=ax,
)
ax.set_title("")
plt.tight_layout()
fig3.savefig(f"{output_dir}/corr_fig.pdf")
fig3.savefig(f"{output_dir}/corr_fig.png", dpi=150)
plt.close(fig3)
print("Figure 3 saved")


# Figure 4: Event Study CAR Plot ----------------------------------------------

# Cumulative abnormal returns over [-5, +5] trading days around earnings
# announcements, split by SUE direction × SameSign.

car_data = (figdata
    .join(
        regdata.select("permno", "rdq", "same_sign", "sue"),
        on=["permno", "rdq"],
        how="inner",
    )
    .with_columns(
        pl.when(pl.col("sue") > 0).then(pl.lit("Positive SUE"))
          .when(pl.col("sue") < 0).then(pl.lit("Negative SUE"))
          .otherwise(pl.lit(None))
          .alias("sue_dir"),
        pl.when(pl.col("same_sign") == 1).then(pl.lit("Same Sign"))
          .otherwise(pl.lit("Diff Sign"))
          .alias("same_sign_lbl"),
    )
    .filter(pl.col("sue_dir").is_not_null())
)

car_summary = (car_data
    .group_by("sue_dir", "same_sign_lbl", "offset")
    .agg(
        pl.col("bhar_cum").mean().alias("mean_bhar"),
        (pl.col("bhar_cum").std() / pl.col("bhar_cum").len().sqrt()).alias("se"),
        pl.len().alias("n"),
    )
    .sort("sue_dir", "same_sign_lbl", "offset")
    .to_pandas()
)

# Create a group column for the interaction
car_summary["group"] = car_summary["sue_dir"] + " / " + car_summary["same_sign_lbl"]

fig4 = (
    ggplot(car_summary, aes(x="offset", y="mean_bhar",
                            color="sue_dir", linetype="same_sign_lbl",
                            group="group"))
    + geom_hline(yintercept=0, linetype="dashed", color="gray")
    + geom_line(size=0.8)
    + geom_point(size=2)
    + scale_x_continuous(
        name="Trading Days Relative to Announcement",
        breaks=range(-5, 6),
    )
    + scale_y_continuous(
        name="Cumulative BHAR",
        labels=lambda x: [f"{v:.1%}" for v in x],
    )
    + scale_color_manual(
        name="Earnings Surprise",
        values={"Positive SUE": "#0051ba", "Negative SUE": "#c41230"},
    )
    + scale_linetype_manual(
        name="SameSign",
        values={"Same Sign": "solid", "Diff Sign": "dashed"},
    )
    + theme_bw(base_family="serif")
    + theme(legend_position="bottom")
)

fig4.save(f"{output_dir}/car_fig.pdf", width=7, height=6)
fig4.save(f"{output_dir}/car_fig.png", width=7, height=6, dpi=150)
print("Figure 4 saved")


# Figure 5: Year-by-Year ERC with Confidence Bands ----------------------------

# The earnings response coefficient (ERC) — slope of BHAR on SUE — estimated
# separately for each year, split by SameSign. Confidence bands show the
# precision of each year's estimate.

reg_pd = regdata.select("fyearq", "same_sign", "bhar", "sue").to_pandas()

# Fit OLS by (fyearq, same_sign) and extract SUE coefficient + CI
erc_rows = []
for (year, ss), grp in reg_pd.groupby(["fyearq", "same_sign"]):
    X = sm.add_constant(grp["sue"])
    model = sm.OLS(grp["bhar"], X).fit()
    ci = model.conf_int().loc["sue"]
    erc_rows.append({
        "fyearq": year,
        "same_sign": ss,
        "estimate": model.params["sue"],
        "conf_low": ci[0],
        "conf_high": ci[1],
    })

ercdata = pd.DataFrame(erc_rows)
ercdata["same_sign_lbl"] = ercdata["same_sign"].map(
    {0: "Different Sign", 1: "Same Sign"}
)

fig5 = (
    ggplot(ercdata, aes(x="fyearq", y="estimate"))
    + geom_ribbon(aes(ymin="conf_low", ymax="conf_high", group="same_sign_lbl"),
                  fill="#CCCCCC", alpha=0.5)
    + geom_line(aes(color="same_sign_lbl"))
    + geom_point(aes(color="same_sign_lbl"))
    + scale_x_continuous(name="Year", breaks=range(1970, 2026, 5))
    + scale_y_continuous(name="ERC (Coefficient on SUE)")
    + scale_color_manual(
        name=" ",
        values={"Same Sign": "#0051ba", "Different Sign": "#c41230"},
    )
    + theme_bw(base_family="serif")
    + theme(legend_position="bottom")
)

fig5.save(f"{output_dir}/erc_year.pdf", width=7, height=6)
fig5.save(f"{output_dir}/erc_year.png", width=7, height=6, dpi=150)
print("Figure 5 saved")

print(f"\nAll figures saved to: {output_dir}")
